File: docs/case-studies/06-authentication-system/java/RefreshTokenHandler.java
// Scenario: refresh token validation in an authentication system
// Demonstrates: session lookup, revocation check, and access token issuance
// Trade-off: session store dependency versus stateless token reuse

package case_studies.auth;

public class RefreshTokenHandler {
    private final SessionStore sessionStore;
    private final TokenService tokenService;

    public RefreshTokenHandler(SessionStore sessionStore, TokenService tokenService) {
        this.sessionStore = sessionStore;
        this.tokenService = tokenService;
    }

    public AuthResponse refresh(RefreshRequest request) {
        Session session = sessionStore.getByRefreshToken(request.getRefreshToken());
        if (session == null || session.isRevoked() || session.isExpired()) {
            throw new UnauthorizedException();
        }

        String accessToken = tokenService.issueAccessToken(session.getUserId());
        return new AuthResponse(accessToken);
    }
}

File: docs/case-studies/06-authentication-system/go/refresh_token_handler.go
// Scenario: refresh token validation in Go
// Demonstrates: session lookup, revocation check, and access token issuance
// Trade-off: session store use versus completely stateless refresh

package auth

import (
	"errors"
	"time"
)

type SessionStore interface {
	GetByRefreshToken(token string) (*Session, error)
}

var ErrUnauthorized = errors.New("unauthorized")

type TokenService interface {
	IssueAccessToken(userID string) (string, error)
}

type RefreshRequest struct {
	RefreshToken string
}

type AuthResponse struct {
	AccessToken string
}

type Session struct {
	UserID    string
	Revoked   bool
	ExpiresAt int64
}

func RefreshToken(req RefreshRequest, store SessionStore, tokenService TokenService) (*AuthResponse, error) {
	session, err := store.GetByRefreshToken(req.RefreshToken)
	if err != nil {
		return nil, err
	}
	if session == nil || session.Revoked || time.Now().Unix() > session.ExpiresAt {
		return nil, ErrUnauthorized
	}

	accessToken, err := tokenService.IssueAccessToken(session.UserID)
	if err != nil {
		return nil, err
	}

	return &AuthResponse{AccessToken: accessToken}, nil
}
